### DISCUSSION TITLE
Error with git dependency when running in github actions

### DISCUSSION THREAD

**Discussion by edulix** [ipetkov/crane#246] [Q&A] [answered]
I have updated my project [flake.nix](https://github.com/sequentech/bulletin-board/blob/main/flake.nix#L43) to use crane and we have a public repo git dependency in `Cargo.toml`.  Within the development environment, `nix build .` works fine. But then, when building the crate in a github action it doesn't:

A)  It either errors saying the following:

```bash
bulletin-board-deps-deps> Caused by:
bulletin-board-deps-deps>   failed to load source for dependency `strand`
bulletin-board-deps-deps> Caused by:
bulletin-board-deps-deps>   Unable to update https://github.com/sequentech/strand
bulletin-board-deps-deps> Caused by:
bulletin-board-deps-deps>   failed to clone into: /build/source/.cargo-home/git/db/strand-0a0cf8eb3b9b64ba
bulletin-board-deps-deps> Caused by:
bulletin-board-deps-deps>   network failure seems to have happened
bulletin-board-deps-deps>   if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
bulletin-board-deps-deps>   https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli
```

(which doesn't happen locally/in-the-vscode-devcontainer).

B) And if I try to allow git fetch  enable CARGO_NET_GIT_FETCH_WITH_CLI="true", then it fails with the following:

```bash
bulletin-board-deps-deps> ++ command cargo check --profile release --all-targets
bulletin-board-deps-deps>     Updating git repository `[https://github.com/sequentech/strand`](https://github.com/sequentech/strand%60)
bulletin-board-deps-deps> error: failed to get `strand` as a dependency of package `bulletin-board v0.1.0 (/build/source)`
bulletin-board-deps-deps> Caused by:
bulletin-board-deps-deps>   failed to load source for dependency `strand`
bulletin-board-deps-deps> Caused by:
bulletin-board-deps-deps>   Unable to update https://github.com/sequentech/strand
bulletin-board-deps-deps> Caused by:
bulletin-board-deps-deps>   failed to clone into: /build/source/.cargo-home/git/db/strand-0a0cf8eb3b9b64ba
bulletin-board-deps-deps> Caused by:
bulletin-board-deps-deps>   could not execute process `git fetch --force --update-head-ok 'https://github.com/sequentech/strand' '+HEAD:refs/remotes/origin/HEAD'` (never executed)
bulletin-board-deps-deps> Caused by:
bulletin-board-deps-deps>   No such file or directory (os error 2)
error: builder for '/nix/store/g73pyrwh5z6dywa1bbl7cvb6m7r4mkq8-bulletin-board-deps-deps-0.1.0.drv' failed with exit code 101;
       last 10 log lines:
       >   Unable to update https://github.com/sequentech/strand
       >
       > Caused by:
       >   failed to clone into: /build/source/.cargo-home/git/db/strand-0a0cf8eb3b9b64ba
       >
       > Caused by:
       >   could not execute process `git fetch --force --update-head-ok 'https://github.com/sequentech/strand' '+HEAD:refs/remotes/origin/HEAD'` (never executed)
       >
       > Caused by:
       >   No such file or directory (os error 2)
```

**Comment by ipetkov** [comment, marked as answer]
Hi @edulix! Have you double checked that the `Cargo.lock` file is up to date and committed to the repo? I took a quick peek but I didn't see the git repo referenced anywhere in the lock file...

(My guess is that when editing locally either rust-analyzer or cargo updated the lock file which makes local builds work, but if the changes aren't committed there is no way for the CI to pick them up)

**Reply by edulix** [reply]
Thanks. The reason was the following:

I was working in a devcontainer in which I thought I might need to also to work in the dependency's code, so I had added strand locally and then forgot about it. The way it works is that I had added a `.cargo/config.toml` file with the following content:

```toml
[patch.'https://github.com/sequentech/strand']
strand = { path = "strand" }
```

Since I was modifying only `Cargo.toml` then calling `cargo update` and I didn't remember this change because in the end I didn't need to change `strand` code, this was not very clear. Anyway, I commented these lines and seems that it works now.
